pkg/newlog/cmd/newlogd.go

```diff
@@ -1000,61 +1098,151 @@ func checkKeepQuota() {
 	}
 
-	totalsize := size0 + size1 + size2 + size3
-	totalCount := len(key0) + len(key1) + len(key2) + len(key3)
+	totalsize := sizeKeepSent + sizeAppUpload + sizeDevUpload + sizeFailSend
+	totalCount := len(filesKeepSent) + len(filesAppUpload) + len(filesDevUpload) + len(fileFailSend)
 	removed := 0
 	// limit file count to not as they can have less size than expected
 	// we can have enormous number of files
 	maxCount := int(maxSize / maxGzipFileSize)
 	if totalsize > maxSize || totalCount > maxCount {
-		keys := key0
-		keys = append(keys, key1...)
-		keys = append(keys, key2...)
-		keys = append(keys, key3...)
-		sort.Strings(keys)
-
-		for _, key := range keys {
-			if _, ok := sfiles[key]; !ok {
-				continue
-			}
-			fs := sfiles[key]
-			filePath := filepath.Join(fs.logDir, fs.filename)
-			if _, err := os.Stat(filePath); err != nil {
-				continue
-			}
-			if err := os.Remove(filePath); err != nil {
-				log.Errorf("checkKeepQuota: remove failed %s, %v", filePath, err)
-				continue
-			}
-			if !fs.isSent {
-				logmetrics.NumGZipFileRemoved++
-			}
-			removed++
-			totalsize -= fs.filesize
-			totalCount--
-			if totalsize < maxSize && totalCount < maxCount {
-				break
+		removalPriority := [][]string{filesKeepSent, fileFailSend, filesAppUpload, filesDevUpload}
+
+		for _, dirFiles := range removalPriority {
+			// sort the files in alphabetical order: this way the files with the oldest (smallest) timestamps will be removed first
+			// side effect: in keepSentQueue, app logs will be removed before device logs, which is okay since those are always synced with the controller
+			sort.Strings(dirFiles)
+
+			for _, filename := range dirFiles {
+				if _, ok := sfiles[filename]; !ok {
+					continue
+				}
+				fs := sfiles[filename]
+				filePath := filepath.Join(fs.logDir, fs.filename)
+				if _, err := os.Stat(filePath); err != nil {
+					continue
+				}
+				if err := os.Remove(filePath); err != nil {
+					log.Errorf("checkKeepQuota: remove failed %s, %v", filePath, err)
+					continue
+				}
+				if fs.logDir == keepSentDir {
+					// since the files are sorted by name and we delete the oldest files first,
+					// we can assume that the latest available log (from the file that is next in line to be deleted)
+					// has the timestamp of the file that was just deleted
+					oldestSavedDeviceLog, err := types.GetTimestampFromGzipName(fs.filename)
+					if err != nil {
+						log.Errorf("checkKeepQuota: %v", err)
+					} else {
+						logmetrics.OldestSavedDeviceLog = oldestSavedDeviceLog
+					}
+				}
+				if !fs.isSent {
+					logmetrics.NumGZipFileRemoved++
+				}
+				removed++
+				totalsize -= fs.filesize
+				totalCount--
+				if totalsize < maxSize && totalCount < maxCount {
+					break
+				}
 			}
 		}
 		log.Tracef("checkKeepQuota: %d gzip files removed", removed)
 	}
+	logmetrics.TotalSizeLogs = uint64(totalsize)
+}
+
+func getLatestLog() (*logs.LogEntry, error) {
```

Comment:
maybe rename this function to getOldestLog()

Reply:
thanks, done!